Subject: [PATCH] videodec: require Cedar picture leases and cedardispsink

Require the native Cedar descriptor ABI; no pixel conversion fallback.
Release picture leases only after display completion.

Upstream-Status: Inappropriate [Miyoo F1C100s private display ABI]

--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -1,3 +1,4 @@
+#include "cedar-display.h"
 /*
  * Copyright (C) 2011, Hewlett-Packard Development Company, L.P.
  *   Author: Sebastian Dröge <sebastian.droege@collabora.co.uk>, Collabora Ltd.
@@ -97,6 +98,25 @@
 G_DEFINE_ABSTRACT_TYPE_WITH_CODE (GstOMXVideoDec, gst_omx_video_dec,
     GST_TYPE_VIDEO_DECODER, DEBUG_INIT);
 
+/* Serialized with downstream buffers, including any queue. The sink stops
+ * DEFE before releasing the last displayed OMX buffer. */
+static void
+cedar_release_display (GstOMXVideoDec *self)
+{
+  gst_pad_push_event (GST_VIDEO_DECODER_SRC_PAD (self),
+        gst_event_new_custom (GST_EVENT_CUSTOM_DOWNSTREAM,
+            gst_structure_new_empty (CEDAR_DISPLAY_RELEASE_EVENT)));
+}
+
+static void
+cedar_output_caps (GstOMXVideoDec *self, GstVideoCodecState *state)
+{
+  gst_caps_replace (&state->caps, NULL);
+  state->caps = gst_video_info_to_caps (&state->info);
+  gst_caps_set_features (state->caps, 0,
+      gst_caps_features_new (CEDAR_DISPLAY_FEATURE, NULL));
+}
+
 static void
 gst_omx_video_dec_class_init (GstOMXVideoDecClass * klass)
 {
@@ -105,6 +125,7 @@
   GstVideoDecoderClass *video_decoder_class = GST_VIDEO_DECODER_CLASS (klass);
 
   gobject_class->finalize = gst_omx_video_dec_finalize;
+
 
   element_class->change_state =
       GST_DEBUG_FUNCPTR (gst_omx_video_dec_change_state);
@@ -125,13 +146,9 @@
 
   klass->cdata.type = GST_OMX_COMPONENT_TYPE_FILTER;
   klass->cdata.default_src_template_caps =
-#if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
-      GST_VIDEO_CAPS_MAKE_WITH_FEATURES (GST_CAPS_FEATURE_MEMORY_GL_MEMORY,
-      "RGBA") "; "
-#endif
-      "video/x-raw, "
-      "width = " GST_VIDEO_SIZE_RANGE ", "
-      "height = " GST_VIDEO_SIZE_RANGE ", " "framerate = " GST_VIDEO_FPS_RANGE;
+      "video/x-raw(" CEDAR_DISPLAY_FEATURE "), format=(string)YV12, "
+      "width=" GST_VIDEO_SIZE_RANGE ", height=" GST_VIDEO_SIZE_RANGE ", "
+      "framerate=" GST_VIDEO_FPS_RANGE;
 }
 
 static void
@@ -199,6 +216,19 @@
   if (!self->dec_in_port || !self->dec_out_port)
     return FALSE;
 
+  {
+    OMX_CONFIG_BOOLEANTYPE enable;
+    OMX_INDEXTYPE index;
+    GST_OMX_INIT_STRUCT (&enable);
+    enable.bEnabled = OMX_TRUE;
+    if (OMX_GetExtensionIndex (self->dec->handle, CEDAR_OMX_EXTENSION, &index) != OMX_ErrorNone ||
+        gst_omx_component_set_parameter (self->dec, index, &enable) != OMX_ErrorNone) {
+      GST_ERROR_OBJECT (self, "libcedarc does not support the Cedar display ABI");
+      return FALSE;
+    }
+    gst_omx_port_update_port_definition (self->dec_out_port, NULL);
+  }
+
   GST_DEBUG_OBJECT (self, "Opened decoder");
 
 #if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
@@ -256,6 +286,7 @@
 {
   OMX_STATETYPE state;
 
+  cedar_release_display (self);
   GST_DEBUG_OBJECT (self, "Shutting down decoder");
 
 #if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
@@ -271,7 +302,8 @@
     gst_omx_component_set_state (self->dec, OMX_StateLoaded);
 
     gst_omx_port_deallocate_buffers (self->dec_in_port);
-    gst_omx_video_dec_deallocate_output_buffers (self);
+    if (gst_omx_video_dec_deallocate_output_buffers (self) != OMX_ErrorNone)
+      return FALSE;
     gst_omx_close_tunnel (self->dec_out_port, self->egl_in_port);
     if (state > OMX_StateLoaded) {
       gst_omx_component_get_state (self->egl_render, 5 * GST_SECOND);
@@ -291,7 +323,8 @@
     }
     gst_omx_component_set_state (self->dec, OMX_StateLoaded);
     gst_omx_port_deallocate_buffers (self->dec_in_port);
-    gst_omx_video_dec_deallocate_output_buffers (self);
+    if (gst_omx_video_dec_deallocate_output_buffers (self) != OMX_ErrorNone)
+      return FALSE;
     if (state > OMX_StateLoaded)
       gst_omx_component_get_state (self->dec, 5 * GST_SECOND);
   }
@@ -550,11 +583,6 @@
 {
   OMX_ERRORTYPE err = OMX_ErrorNone;
   GstOMXPort *port;
-  GstBufferPool *pool;
-  GstStructure *config;
-  gboolean eglimage = FALSE, add_videometa = FALSE;
-  GstCaps *caps = NULL;
-  guint min = 0, max = 0;
   GstVideoCodecState *state =
       gst_video_decoder_get_output_state (GST_VIDEO_DECODER (self));
 
@@ -564,342 +592,62 @@
   port = self->dec_out_port;
 #endif
 
-  pool = gst_video_decoder_get_buffer_pool (GST_VIDEO_DECODER (self));
-  if (pool) {
-    GstAllocator *allocator;
-
-    config = gst_buffer_pool_get_config (pool);
-    if (!gst_buffer_pool_config_get_params (config, &caps, NULL, &min, &max)) {
-      GST_ERROR_OBJECT (self, "Can't get buffer pool params");
-      gst_structure_free (config);
-      err = OMX_ErrorUndefined;
-      goto done;
-    }
-    if (!gst_buffer_pool_config_get_allocator (config, &allocator, NULL)) {
-      GST_ERROR_OBJECT (self, "Can't get buffer pool allocator");
-      gst_structure_free (config);
-      err = OMX_ErrorUndefined;
-      goto done;
-    }
-
-    /* Need at least 2 buffers for anything meaningful */
-    min = MAX (MAX (min, port->port_def.nBufferCountMin), 4);
-    if (max == 0) {
-      max = min;
-    } else if (max < port->port_def.nBufferCountMin || max < 2) {
-      /* Can't use pool because can't have enough buffers */
-      gst_caps_replace (&caps, NULL);
-    } else {
-      min = max;
-    }
-
-    add_videometa = gst_buffer_pool_config_has_option (config,
-        GST_BUFFER_POOL_OPTION_VIDEO_META);
-    gst_structure_free (config);
-
-#if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
-    eglimage = self->eglimage
-        && (allocator && GST_IS_GL_MEMORY_EGL_ALLOCATOR (allocator));
-#else
-    /* TODO: Implement something that works for other targets too */
-    eglimage = FALSE;
-#endif
-    caps = caps ? gst_caps_ref (caps) : NULL;
-
-    GST_DEBUG_OBJECT (self, "Trying to use pool %p with caps %" GST_PTR_FORMAT
-        " and memory type %s", pool, caps,
-        (allocator ? allocator->mem_type : "(null)"));
-  } else {
-    gst_caps_replace (&caps, NULL);
-    min = max = port->port_def.nBufferCountMin;
-    GST_DEBUG_OBJECT (self, "No pool available, not negotiated yet");
-  }
-
-#if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
-  /* Will retry without EGLImage */
-  if (self->eglimage && !eglimage) {
-    GST_DEBUG_OBJECT (self,
-        "Wanted to use EGLImage but downstream doesn't support it");
-    err = OMX_ErrorUndefined;
-    goto done;
-  }
-#endif
-
-  if (caps)
-    self->out_port_pool =
-        gst_omx_buffer_pool_new (GST_ELEMENT_CAST (self), self->dec, port);
-
-#if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
-  if (eglimage) {
-    GList *buffers = NULL;
-    GList *images = NULL;
-    gint i;
-    GstBufferPoolAcquireParams params = { 0, };
-    EGLDisplay egl_display = EGL_NO_DISPLAY;
-
-    GST_DEBUG_OBJECT (self, "Trying to allocate %d EGLImages", min);
-
-    for (i = 0; i < min; i++) {
-      GstBuffer *buffer;
-      GstMemory *mem;
-      GstGLMemoryEGL *gl_mem;
-
-      if (gst_buffer_pool_acquire_buffer (pool, &buffer, &params) != GST_FLOW_OK
-          || gst_buffer_n_memory (buffer) != 1
-          || !(mem = gst_buffer_peek_memory (buffer, 0))
-          || !GST_IS_GL_MEMORY_EGL_ALLOCATOR (mem->allocator)) {
-        GST_INFO_OBJECT (self, "Failed to allocated %d-th EGLImage", i);
-        g_list_free_full (buffers, (GDestroyNotify) gst_buffer_unref);
-        g_list_free (images);
-        buffers = NULL;
-        images = NULL;
-        /* TODO: For non-RPi targets we want to use the normal memory code below */
-        /* Retry without EGLImage */
-        err = OMX_ErrorUndefined;
-        goto done;
-      }
-      gl_mem = (GstGLMemoryEGL *) mem;
-      buffers = g_list_append (buffers, buffer);
-      /* FIXME: replace with the affine transformation meta */
-      gl_mem->image->orientation =
-          GST_VIDEO_GL_TEXTURE_ORIENTATION_X_NORMAL_Y_FLIP;
-      images = g_list_append (images, gst_gl_memory_egl_get_image (gl_mem));
-      if (egl_display == EGL_NO_DISPLAY)
-        egl_display = gst_gl_memory_egl_get_display (gl_mem);
-    }
-
-    GST_DEBUG_OBJECT (self, "Allocated %d EGLImages successfully", min);
-
-    /* Everything went fine? */
-    if (eglimage) {
-      GST_DEBUG_OBJECT (self, "Setting EGLDisplay");
-      self->egl_out_port->port_def.format.video.pNativeWindow = egl_display;
-      err =
-          gst_omx_port_update_port_definition (self->egl_out_port,
-          &self->egl_out_port->port_def);
-      if (err != OMX_ErrorNone) {
-        GST_INFO_OBJECT (self,
-            "Failed to set EGLDisplay on port: %s (0x%08x)",
-            gst_omx_error_to_string (err), err);
-        g_list_free_full (buffers, (GDestroyNotify) gst_buffer_unref);
-        g_list_free (images);
-        /* TODO: For non-RPi targets we want to use the normal memory code below */
-        /* Retry without EGLImage */
-        goto done;
-      } else {
-        GList *l;
-
-        if (min != port->port_def.nBufferCountActual) {
-          err = gst_omx_port_update_port_definition (port, NULL);
-          if (err == OMX_ErrorNone) {
-            port->port_def.nBufferCountActual = min;
-            err = gst_omx_port_update_port_definition (port, &port->port_def);
-          }
-
-          if (err != OMX_ErrorNone) {
-            GST_INFO_OBJECT (self,
-                "Failed to configure %u output buffers: %s (0x%08x)", min,
-                gst_omx_error_to_string (err), err);
-            g_list_free_full (buffers, (GDestroyNotify) gst_buffer_unref);
-            g_list_free (images);
-            /* TODO: For non-RPi targets we want to use the normal memory code below */
-            /* Retry without EGLImage */
-
-            goto done;
-          }
-        }
-
-        if (!gst_omx_port_is_enabled (port)) {
-          err = gst_omx_port_set_enabled (port, TRUE);
-          if (err != OMX_ErrorNone) {
-            GST_INFO_OBJECT (self,
-                "Failed to enable port: %s (0x%08x)",
-                gst_omx_error_to_string (err), err);
-            g_list_free_full (buffers, (GDestroyNotify) gst_buffer_unref);
-            g_list_free (images);
-            /* TODO: For non-RPi targets we want to use the normal memory code below */
-            /* Retry without EGLImage */
-            goto done;
-          }
-        }
-
-        err = gst_omx_port_use_eglimages (port, images);
-        g_list_free (images);
-
-        if (err != OMX_ErrorNone) {
-          GST_INFO_OBJECT (self,
-              "Failed to pass EGLImages to port: %s (0x%08x)",
-              gst_omx_error_to_string (err), err);
-          g_list_free_full (buffers, (GDestroyNotify) gst_buffer_unref);
-          /* TODO: For non-RPi targets we want to use the normal memory code below */
-          /* Retry without EGLImage */
-          goto done;
-        }
-
-        err = gst_omx_port_wait_enabled (port, 2 * GST_SECOND);
-        if (err != OMX_ErrorNone) {
-          GST_INFO_OBJECT (self,
-              "Failed to wait until port is enabled: %s (0x%08x)",
-              gst_omx_error_to_string (err), err);
-          g_list_free_full (buffers, (GDestroyNotify) gst_buffer_unref);
-          /* TODO: For non-RPi targets we want to use the normal memory code below */
-          /* Retry without EGLImage */
-          goto done;
-        }
-
-        GST_DEBUG_OBJECT (self, "Populating internal buffer pool");
-        GST_OMX_BUFFER_POOL (self->out_port_pool)->other_pool =
-            GST_BUFFER_POOL (gst_object_ref (pool));
-        for (l = buffers; l; l = l->next) {
-          g_ptr_array_add (GST_OMX_BUFFER_POOL (self->out_port_pool)->buffers,
-              l->data);
-        }
-        g_list_free (buffers);
-        /* All good and done, set caps below */
-      }
-    }
-  }
-#endif
-
-  /* If not using EGLImage or trying to use EGLImage failed */
-  if (!eglimage) {
-    gboolean was_enabled = TRUE;
-
-    if (min != port->port_def.nBufferCountActual) {
-      err = gst_omx_port_update_port_definition (port, NULL);
-      if (err == OMX_ErrorNone) {
-        port->port_def.nBufferCountActual = min;
-        err = gst_omx_port_update_port_definition (port, &port->port_def);
-      }
-
-      if (err != OMX_ErrorNone) {
-        GST_ERROR_OBJECT (self,
-            "Failed to configure %u output buffers: %s (0x%08x)", min,
-            gst_omx_error_to_string (err), err);
-        goto done;
-      }
-    }
-
-    if (!gst_omx_port_is_enabled (port)) {
+  {
+    GstStructure *cedar_config;
+    gboolean was_enabled = gst_omx_port_is_enabled (port);
+    if (!state || !state->caps || port->port_def.nBufferSize != sizeof(struct cedar_display_frame)) {
+      GST_ERROR_OBJECT (self, "Unexpected Cedar descriptor size or missing output caps");
+      if (state) gst_video_codec_state_unref (state);
+      return OMX_ErrorBadParameter;
+    }
+    if (!was_enabled) {
       err = gst_omx_port_set_enabled (port, TRUE);
       if (err != OMX_ErrorNone) {
-        GST_INFO_OBJECT (self,
-            "Failed to enable port: %s (0x%08x)",
-            gst_omx_error_to_string (err), err);
-        goto done;
-      }
-      was_enabled = FALSE;
-    }
-
-    err = gst_omx_port_allocate_buffers (port);
-    if (err != OMX_ErrorNone && min > port->port_def.nBufferCountMin) {
-      GST_ERROR_OBJECT (self,
-          "Failed to allocate required number of buffers %d, trying less and copying",
-          min);
-      min = port->port_def.nBufferCountMin;
-
-      if (!was_enabled) {
-        err = gst_omx_port_set_enabled (port, FALSE);
-        if (err != OMX_ErrorNone) {
-          GST_INFO_OBJECT (self,
-              "Failed to disable port again: %s (0x%08x)",
-              gst_omx_error_to_string (err), err);
-          goto done;
-        }
-      }
-
-      if (min != port->port_def.nBufferCountActual) {
-        err = gst_omx_port_update_port_definition (port, NULL);
-        if (err == OMX_ErrorNone) {
-          port->port_def.nBufferCountActual = min;
-          err = gst_omx_port_update_port_definition (port, &port->port_def);
-        }
-
-        if (err != OMX_ErrorNone) {
-          GST_ERROR_OBJECT (self,
-              "Failed to configure %u output buffers: %s (0x%08x)", min,
-              gst_omx_error_to_string (err), err);
-          goto done;
-        }
-      }
-
+        if (state) gst_video_codec_state_unref (state);
+        return err;
+      }
+    }
+    /* no-disable-outport components allocate OMX headers before the first
+     * output caps. In that case only wrap the existing allocations. */
+    if (!port->buffers || !port->buffers->len)
       err = gst_omx_port_allocate_buffers (port);
-
-      /* Can't provide buffers downstream in this case */
-      gst_caps_replace (&caps, NULL);
-    }
-
+    if (err == OMX_ErrorNone && !was_enabled)
+      err = gst_omx_port_wait_enabled (port, 2 * GST_SECOND);
     if (err != OMX_ErrorNone) {
-      GST_ERROR_OBJECT (self, "Failed to allocate %d buffers: %s (0x%08x)", min,
-          gst_omx_error_to_string (err), err);
-      goto done;
-    }
-
-    if (!was_enabled) {
-      err = gst_omx_port_wait_enabled (port, 2 * GST_SECOND);
-      if (err != OMX_ErrorNone) {
-        GST_ERROR_OBJECT (self,
-            "Failed to wait until port is enabled: %s (0x%08x)",
-            gst_omx_error_to_string (err), err);
-        goto done;
-      }
-    }
-
-
-  }
-
-  err = OMX_ErrorNone;
-
-  if (caps) {
-    config = gst_buffer_pool_get_config (self->out_port_pool);
-
-    if (add_videometa)
-      gst_buffer_pool_config_add_option (config,
-          GST_BUFFER_POOL_OPTION_VIDEO_META);
-
-    gst_buffer_pool_config_set_params (config, caps,
-        self->dec_out_port->port_def.nBufferSize, min, max);
-
-    if (!gst_buffer_pool_set_config (self->out_port_pool, config)) {
-      GST_INFO_OBJECT (self, "Failed to set config on internal pool");
-      gst_object_unref (self->out_port_pool);
-      self->out_port_pool = NULL;
-      goto done;
-    }
-
-    GST_OMX_BUFFER_POOL (self->out_port_pool)->allocating = TRUE;
-    /* This now allocates all the buffers */
-    if (!gst_buffer_pool_set_active (self->out_port_pool, TRUE)) {
-      GST_INFO_OBJECT (self, "Failed to activate internal pool");
-      gst_object_unref (self->out_port_pool);
-      self->out_port_pool = NULL;
-    } else {
+      if (state) gst_video_codec_state_unref (state);
+      return err;
+    }
+    self->out_port_pool = gst_omx_buffer_pool_new (GST_ELEMENT_CAST (self), self->dec, port);
+    cedar_config = gst_buffer_pool_get_config (self->out_port_pool);
+    gst_buffer_pool_config_set_params (cedar_config, state->caps,
+        port->port_def.nBufferSize, port->port_def.nBufferCountActual,
+        port->port_def.nBufferCountActual);
+    if (!gst_buffer_pool_set_config (self->out_port_pool, cedar_config))
+      err = OMX_ErrorUndefined;
+    else {
+      GST_OMX_BUFFER_POOL (self->out_port_pool)->allocating = TRUE;
+      if (!gst_buffer_pool_set_active (self->out_port_pool, TRUE))
+        err = OMX_ErrorUndefined;
       GST_OMX_BUFFER_POOL (self->out_port_pool)->allocating = FALSE;
     }
-  } else if (self->out_port_pool) {
-    gst_object_unref (self->out_port_pool);
-    self->out_port_pool = NULL;
-  }
-
-done:
-  if (!self->out_port_pool && err == OMX_ErrorNone)
-    GST_DEBUG_OBJECT (self,
-        "Not using our internal pool and copying buffers for downstream");
-
-  if (caps)
-    gst_caps_unref (caps);
-  if (pool)
-    gst_object_unref (pool);
-  if (state)
     gst_video_codec_state_unref (state);
-
-  return err;
+    return err;
+  }
+
 }
 
 static OMX_ERRORTYPE
 gst_omx_video_dec_deallocate_output_buffers (GstOMXVideoDec * self)
 {
   OMX_ERRORTYPE err;
+
+  if (self->out_port_pool) {
+    /* The stock pool teardown does not wait for downstream users. Never free
+     * OMX descriptors while a sink/queue still owns a picture lease. */
+    err = gst_omx_port_wait_buffers_released (self->dec_out_port, 5 * GST_SECOND);
+    if (err != OMX_ErrorNone)
+      return err;
+  }
 
   if (self->out_port_pool) {
     gst_buffer_pool_set_active (self->out_port_pool, FALSE);
@@ -1139,6 +887,7 @@
   state = gst_video_decoder_set_output_state (GST_VIDEO_DECODER (self),
       format, port_def.format.video.nFrameWidth,
       port_def.format.video.nFrameHeight, self->input_state);
+  cedar_output_caps (self, state);
 
   if (!gst_video_decoder_negotiate (GST_VIDEO_DECODER (self))) {
     gst_video_codec_state_unref (state);
@@ -1280,6 +1029,7 @@
     /* Reallocate all buffers */
     if (acq_return == GST_OMX_ACQUIRE_BUFFER_RECONFIGURE
         && gst_omx_port_is_enabled (port)) {
+      cedar_release_display (self);
       err = gst_omx_port_set_enabled (port, FALSE);
       if (err != OMX_ErrorNone)
         goto reconfigure_error;
@@ -1334,6 +1084,8 @@
           format, port_def.format.video.nFrameWidth,
           port_def.format.video.nFrameHeight, self->input_state);
 
+      cedar_output_caps (self, state);
+
       /* Take framerate and pixel-aspect-ratio from sinkpad caps */
 
       if (!gst_video_decoder_negotiate (GST_VIDEO_DECODER (self))) {
@@ -1346,6 +1098,17 @@
       gst_video_codec_state_unref (state);
 
       GST_VIDEO_DECODER_STREAM_UNLOCK (self);
+    }
+
+    if (!self->out_port_pool) {
+      err = gst_omx_video_dec_allocate_output_buffers (self);
+      if (err != OMX_ErrorNone) {
+        if (buf) {
+          gst_omx_port_release_buffer (port, buf);
+          buf = NULL;
+        }
+        goto reconfigure_error;
+      }
     }
 
     /* Now get a buffer */
@@ -1685,6 +1448,7 @@
 
   self = GST_OMX_VIDEO_DEC (decoder);
 
+  cedar_release_display (self);
   GST_DEBUG_OBJECT (self, "Stopping decoder");
 
   gst_omx_port_set_flushing (self->dec_in_port, 5 * GST_SECOND, TRUE);
@@ -1732,93 +1496,12 @@
 gst_omx_video_dec_negotiate (GstOMXVideoDec * self)
 {
   OMX_VIDEO_PARAM_PORTFORMATTYPE param;
-  OMX_ERRORTYPE err;
-  GstCaps *comp_supported_caps;
-  GList *negotiation_map = NULL, *l;
-  GstCaps *templ_caps, *intersection;
-  GstVideoFormat format;
-  GstStructure *s;
-  const gchar *format_str;
-
-  GST_DEBUG_OBJECT (self, "Trying to negotiate a video format with downstream");
-
-  templ_caps = gst_pad_get_pad_template_caps (GST_VIDEO_DECODER_SRC_PAD (self));
-  intersection =
-      gst_pad_peer_query_caps (GST_VIDEO_DECODER_SRC_PAD (self), templ_caps);
-  gst_caps_unref (templ_caps);
-
-  GST_DEBUG_OBJECT (self, "Allowed downstream caps: %" GST_PTR_FORMAT,
-      intersection);
-
-  negotiation_map =
-      gst_omx_video_get_supported_colorformats (self->dec_out_port,
-      self->input_state);
-
-  comp_supported_caps = gst_omx_video_get_caps_for_map (negotiation_map);
-
-  if (!gst_caps_is_empty (comp_supported_caps)) {
-    GstCaps *tmp;
-
-    tmp = gst_caps_intersect (comp_supported_caps, intersection);
-    gst_caps_unref (intersection);
-    intersection = tmp;
-  }
-  gst_caps_unref (comp_supported_caps);
-
-  if (gst_caps_is_empty (intersection)) {
-    gst_caps_unref (intersection);
-    GST_ERROR_OBJECT (self, "Empty caps");
-    g_list_free_full (negotiation_map,
-        (GDestroyNotify) gst_omx_video_negotiation_map_free);
-    return FALSE;
-  }
-
-  intersection = gst_caps_truncate (intersection);
-  intersection = gst_caps_fixate (intersection);
-
-  s = gst_caps_get_structure (intersection, 0);
-  format_str = gst_structure_get_string (s, "format");
-  if (!format_str ||
-      (format =
-          gst_video_format_from_string (format_str)) ==
-      GST_VIDEO_FORMAT_UNKNOWN) {
-    GST_ERROR_OBJECT (self, "Invalid caps: %" GST_PTR_FORMAT, intersection);
-    gst_caps_unref (intersection);
-    g_list_free_full (negotiation_map,
-        (GDestroyNotify) gst_omx_video_negotiation_map_free);
-    return FALSE;
-  }
-
   GST_OMX_INIT_STRUCT (&param);
   param.nPortIndex = self->dec_out_port->index;
-
-  for (l = negotiation_map; l; l = l->next) {
-    GstOMXVideoNegotiationMap *m = l->data;
-
-    if (m->format == format) {
-      param.eColorFormat = m->type;
-      break;
-    }
-  }
-
-  GST_DEBUG_OBJECT (self, "Negotiating color format %s (%d)", format_str,
-      param.eColorFormat);
-
-  /* We must find something here */
-  g_assert (l != NULL);
-  g_list_free_full (negotiation_map,
-      (GDestroyNotify) gst_omx_video_negotiation_map_free);
-
-  err =
-      gst_omx_component_set_parameter (self->dec,
-      OMX_IndexParamVideoPortFormat, &param);
-  if (err != OMX_ErrorNone) {
-    GST_ERROR_OBJECT (self, "Failed to set video port format: %s (0x%08x)",
-        gst_omx_error_to_string (err), err);
-  }
-
-  gst_caps_unref (intersection);
-  return (err == OMX_ErrorNone);
+  param.eColorFormat = OMX_COLOR_FormatYVU420Planar;
+  param.eCompressionFormat = OMX_VIDEO_CodingUnused;
+  return gst_omx_component_set_parameter (self->dec,
+      OMX_IndexParamVideoPortFormat, &param) == OMX_ErrorNone;
 }
 
 static gboolean
@@ -2542,68 +2225,32 @@
 static gboolean
 gst_omx_video_dec_decide_allocation (GstVideoDecoder * bdec, GstQuery * query)
 {
+  GstCaps *caps;
   GstBufferPool *pool;
   GstStructure *config;
 
-#if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_GL)
-  {
-    GstCaps *caps;
-    gint i, n;
-    GstVideoInfo info;
-
-    gst_query_parse_allocation (query, &caps, NULL);
-    if (caps && gst_video_info_from_caps (&info, caps)
-        && info.finfo->format == GST_VIDEO_FORMAT_RGBA) {
-      gboolean found = FALSE;
-      GstCapsFeatures *feature = gst_caps_get_features (caps, 0);
-      /* Prefer an EGLImage allocator if available and we want to use it */
-      n = gst_query_get_n_allocation_params (query);
-      for (i = 0; i < n; i++) {
-        GstAllocator *allocator;
-        GstAllocationParams params;
-
-        gst_query_parse_nth_allocation_param (query, i, &allocator, &params);
-        if (allocator) {
-          if (GST_IS_GL_MEMORY_EGL_ALLOCATOR (allocator)) {
-            found = TRUE;
-            gst_query_set_nth_allocation_param (query, 0, allocator, &params);
-            while (gst_query_get_n_allocation_params (query) > 1)
-              gst_query_remove_nth_allocation_param (query, 1);
-          }
-
-          gst_object_unref (allocator);
-
-          if (found)
-            break;
-        }
-      }
-
-      /* if try to negotiate with caps feature memory:EGLImage
-       * and if allocator is not of type memory EGLImage then fails */
-      if (feature
-          && gst_caps_features_contains (feature,
-              GST_CAPS_FEATURE_MEMORY_GL_MEMORY) && !found) {
-        return FALSE;
-      }
-    }
-  }
-#endif
-
-  if (!GST_VIDEO_DECODER_CLASS
-      (gst_omx_video_dec_parent_class)->decide_allocation (bdec, query))
+  gst_query_parse_allocation (query, &caps, NULL);
+  if (!caps || gst_caps_is_empty (caps) ||
+      !gst_caps_features_contains (gst_caps_get_features (caps, 0),
+          CEDAR_DISPLAY_FEATURE))
     return FALSE;
 
-  g_assert (gst_query_get_n_allocation_pools (query) > 0);
-  gst_query_parse_nth_allocation_pool (query, 0, &pool, NULL, NULL, NULL);
-  g_assert (pool != NULL);
-
+  /* GstVideoDecoder 1.20 requires a non-NULL pool in the ALLOCATION reply.
+   * This lazy pool describes only the descriptor-sized output ABI. Actual
+   * output comes from out_port_pool, after OMX port setup/reconfiguration;
+   * never use GstVideoBufferPool, which would allocate full pixel frames. */
+  pool = gst_buffer_pool_new ();
   config = gst_buffer_pool_get_config (pool);
-  if (gst_query_find_allocation_meta (query, GST_VIDEO_META_API_TYPE, NULL)) {
-    gst_buffer_pool_config_add_option (config,
-        GST_BUFFER_POOL_OPTION_VIDEO_META);
-  }
-  gst_buffer_pool_set_config (pool, config);
+  gst_buffer_pool_config_set_params (config, caps,
+      sizeof (struct cedar_display_frame), 0, 0);
+  if (!gst_buffer_pool_set_config (pool, config)) {
+    gst_object_unref (pool);
+    return FALSE;
+  }
+  while (gst_query_get_n_allocation_pools (query))
+    gst_query_remove_nth_allocation_pool (query, 0);
+  gst_query_add_allocation_pool (query, pool,
+      sizeof (struct cedar_display_frame), 0, 0);
   gst_object_unref (pool);
-
   return TRUE;
 }
--- a/omx/gstomxbufferpool.c
+++ b/omx/gstomxbufferpool.c
@@ -1,3 +1,4 @@
+#include "cedar-display.h"
 /*
  * Copyright (C) 2011, Hewlett-Packard Development Company, L.P.
  *   Author: Sebastian Dröge <sebastian.droege@collabora.co.uk>, Collabora Ltd.
@@ -382,6 +383,13 @@
       }
     }
 
+    pool->need_copy = FALSE;
+  } else if (gst_caps_features_contains (gst_caps_get_features (pool->caps, 0),
+          CEDAR_DISPLAY_FEATURE)) {
+    GstMemory *mem = gst_omx_memory_allocator_alloc (pool->allocator, 0, omx_buf);
+    buf = gst_buffer_new ();
+    gst_buffer_append_memory (buf, mem);
+    g_ptr_array_add (pool->buffers, buf);
     pool->need_copy = FALSE;
   } else {
     GstMemory *mem;
@@ -539,6 +547,10 @@
     omx_buf =
         gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (buffer),
         gst_omx_buffer_data_quark);
+    if (pool->caps && !omx_buf->used &&
+        gst_caps_features_contains (gst_caps_get_features (pool->caps, 0), CEDAR_DISPLAY_FEATURE))
+      gst_omx_component_set_config (pool->component,
+          (OMX_INDEXTYPE)CEDAR_OMX_RELEASE_INDEX, omx_buf->omx_buf);
     if (pool->port->port_def.eDir == OMX_DirOutput && !omx_buf->used) {
       /* Release back to the port, can be filled again */
       err = gst_omx_port_release_buffer (pool->port, omx_buf);
